sdks/java/io/rabbitmq/src/test/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIOTest.java

```diff
@@ -185,20 +234,154 @@ public void testReadExchange() throws Exception {
+  public void testReadDeclaredTopicExchangeWithQueueDeclare() throws Exception {
+    doExchangeTest(
+        new ExchangeTestPlan(
+            RabbitMqIO.read()
+                .withExchange("DeclaredTopicExchangeWithQueueDeclare", "topic", "#")
+                .withQueue("declared-queue")
+                .withQueueDeclare(true),
+            10));
+  }
+
+  @Test(timeout = ONE_MINUTE_MS)
+  public void testReadDeclaredTopicExchange() throws Exception {
+    final int numRecords = 10;
+    RabbitMqIO.Read read =
+        RabbitMqIO.read().withExchange("DeclaredTopicExchange", "topic", "user.create.#");
+
+    final Supplier<String> publishRoutingKeyGen =
+        new Supplier<String>() {
+          private AtomicInteger counter = new AtomicInteger(0);
+
+          @Override
+          public String get() {
+            int count = counter.getAndIncrement();
+            if (count % 2 == 0) {
+              return "user.create." + count;
+            }
+            return "user.delete." + count;
+          }
+        };
+
+    ExchangeTestPlan plan =
+        new ExchangeTestPlan(read, numRecords / 2, numRecords) {
+          @Override
+          public Supplier<String> publishRoutingKeyGen() {
+            return publishRoutingKeyGen;
+          }
+
+          @Override
+          public List<String> expectedResults() {
+            return IntStream.range(0, numRecords)
+                .filter(i -> i % 2 == 0)
+                .mapToObj(RabbitMqTestUtils::generateRecord)
+                .map(RabbitMqTestUtils::recordToString)
+                .collect(Collectors.toList());
+          }
+        };
+
+    doExchangeTest(plan);
+  }
+
+  @Test(timeout = ONE_MINUTE_MS)
+  public void testDeclareIncompatibleExchangeFails() throws Exception {
+    RabbitMqIO.Read read =
+        RabbitMqIO.read().withExchange("IncompatibleExchange", "direct", "unused");
+    try {
+      doExchangeTest(new ExchangeTestPlan(read, 1), true);
+      fail("Expected to have failed to declare an incompatible exchange");
+    } catch (Exception e) {
+      Throwable cause = Throwables.getRootCause(e);
+      if (cause instanceof ShutdownSignalException) {
+        ShutdownSignalException sse = (ShutdownSignalException) cause;
+        Method reason = sse.getReason();
+        if (reason instanceof com.rabbitmq.client.AMQP.Connection.Close) {
+          com.rabbitmq.client.AMQP.Connection.Close close =
+              (com.rabbitmq.client.AMQP.Connection.Close) reason;
+          assertEquals("Expected failure is 530: not-allowed", 530, close.getReplyCode());
+        } else {
+          fail(
+              "Unexpected ShutdownSignalException reason. Expected Connection.Close. Got: "
+                  + reason);
+        }
+      } else {
+        fail("Expected to fail with ShutdownSignalException. Instead failed with " + cause);
+      }
+    }
+  }
+
+  @Test(timeout = ONE_MINUTE_MS)
+  public void testUseCorrelationIdSucceedsWhenIdsPresent() throws Exception {
```

Comment:
This test and the one below retroactively add testing support for https://github.com/apache/beam/pull/9782 (BEAM-8390) now that it's easier to write this test.